File: hello-data-portal/hello-data-portal-api/src/main/java/ch/bedag/dap/hellodata/portal/superset/pdfexport/PdfExportService.java
/*
 * Copyright © 2024, Kanton Bern
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the <organization> nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package ch.bedag.dap.hellodata.portal.superset.pdfexport;

import ch.bedag.dap.hellodata.commons.security.SecurityUtils;
import ch.bedag.dap.hellodata.commons.sidecars.resources.v1.dashboard.screenshot.ChartScreenshotRequest.ChartSpec;
import ch.bedag.dap.hellodata.portal.superset.service.DashboardService;
import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

/** Orchestrates the custom-layout PDF export: authZ gate, chart sizing, screenshot fetch over
 *  NATS, grid assembly and PDF rendering. */
@Log4j2
@Service
@RequiredArgsConstructor
public class PdfExportService {

    private static final Duration SCREENSHOT_TIMEOUT = Duration.ofSeconds(120);
    private static final int SCREENSHOT_DPI = ReportTemplate.DEFAULT_SCREENSHOT_DPI;

    private final DashboardService dashboardService;
    private final ScreenshotClient screenshotClient;
    private final PdfRenderer pdfRenderer;

    /** Custom grid export designed in the Angular builder. */
    public byte[] exportCustom(PdfLayoutRequest request) {
        dashboardService.assertCurrentUserMayAccess(request.instanceName(), request.dashboardId());
        if (request.items() == null || request.items().isEmpty()) {
            throw new IllegalStateException("Layout has no cells to export");
        }
        ReportTemplate template = ReportTemplate.fromId(request.template());

        // One screenshot per distinct (chart, cols, rows); specId disambiguates the same chart at two sizes.
        Map<String, ChartSpec> specsById = new LinkedHashMap<>();
        for (PdfLayoutRequest.Item item : request.items()) {
            if (!item.isChart()) {
                continue;
            }
            int cols = LayoutGridPacker.clampCols(item.cols());
            int rows = LayoutGridPacker.clampRowsToBand(item.y(), item.rows());
            String specId = LayoutGridPacker.specId(item.chartId(), cols, rows);
            int[] size = template.chartScreenshotPx(cols, rows, SCREENSHOT_DPI);
            specsById.putIfAbsent(specId, new ChartSpec(specId, item.chartId(), size[0], size[1]));
        }

        // Render as the requesting user so Superset applies that user's row-level-security filters.
        String userEmail = SecurityUtils.getCurrentUserEmail();
        Map<String, byte[]> pngs = screenshotClient.fetchScreenshots(request.instanceName(), userEmail, new ArrayList<>(specsById.values()), SCREENSHOT_TIMEOUT);

        CustomLayout layout = LayoutGridPacker.buildCustomLayout(request, pngs);
        return pdfRenderer.renderCustom(layout, template);
    }
}

File: hello-data-portal/hello-data-portal-ui/src/app/store/pdf-export/pdf-export.service.ts
///
/// Copyright © 2024, Kanton Bern
/// All rights reserved.
///
/// Redistribution and use in source and binary forms, with or without
/// modification, are permitted provided that the following conditions are met:
///     * Redistributions of source code must retain the above copyright
///       notice, this list of conditions and the following disclaimer.
///     * Redistributions in binary form must reproduce the above copyright
///       notice, this list of conditions and the following disclaimer in the
///       documentation and/or other materials provided with the distribution.
///     * Neither the name of the <organization> nor the
///       names of its contributors may be used to endorse or promote products
///       derived from this software without specific prior written permission.
///
/// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
/// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
/// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
/// DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
/// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
/// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
/// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
/// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
/// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
/// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
///

import {inject, Injectable} from "@angular/core";
import {HttpClient, HttpParams} from "@angular/common/http";
import {Observable} from "rxjs";
import {environment} from "../../../environments/environment";
import {PdfChartRef, PdfLayoutRequest} from "./pdf-export.model";

@Injectable({providedIn: 'root'})
export class PdfExportService {
  private httpClient = inject(HttpClient);

  private baseUrl = `${environment.portalApi}/superset/dashboards`;

  /** Charts of one dashboard for the builder palette. */
  public getCharts(instanceName: string, dashboardId: number): Observable<PdfChartRef[]> {
    return this.httpClient.get<PdfChartRef[]>(
      `${this.baseUrl}/${encodeURIComponent(instanceName)}/${dashboardId}/charts`
    );
  }

  /** Existing markdown/text blocks of one dashboard for the builder palette. */
  public getMarkdownBlocks(instanceName: string, dashboardId: number): Observable<string[]> {
    return this.httpClient.get<string[]>(
      `${this.baseUrl}/${encodeURIComponent(instanceName)}/${dashboardId}/markdown`
    );
  }

  /** Render the custom grid layout to a PDF (streamed as application/pdf). */
  public exportCustom(request: PdfLayoutRequest): Observable<Blob> {
    return this.httpClient.post(`${this.baseUrl}/pdf/custom`, request, {responseType: 'blob'});
  }

  /** Single-chart screenshot preview (PNG blob), sized to the tile's span, for the builder grid. */
  public getChartPreview(instanceName: string, dashboardId: number, chartId: number,
                         cols: number, rows: number, template: string): Observable<Blob> {
    const params = new HttpParams().set('cols', cols).set('rows', rows).set('template', template);
    return this.httpClient.get(
      `${this.baseUrl}/${encodeURIComponent(instanceName)}/${dashboardId}/charts/${chartId}/preview`,
      {params, responseType: 'blob'}
    );
  }
}
